File: py/aoc2024/day2.py
"""
Day 2: Red-Nosed Reports
"""

from typing import Generator

SAMPLE_INPUT = """
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
"""


def _parse(data: str) -> Generator[list[int]]:
    for line in data.splitlines():
        line = line.strip()
        if line:
            yield [int(level) for level in line.split()]


def _issafe1(report: list[int]) -> bool:
    decreasing, increasing = False, False
    for x, y in zip(report, report[1:]):
        delta = x - y
        if -3 <= delta <= -1:
            decreasing = True
            if increasing:
                break
        elif 1 <= delta <= 3:
            increasing = True
            if decreasing:
                break
        else:
            break
    else:
        return True
    return False


def _issafe2(report: list[int]) -> bool:
    report2 = report[:-1].copy()
    for i in range(len(report2), 0, -1):
        if _issafe1(report2):
            return True
        report2[i - 1] = report[i]
    return _issafe1(report2)


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT)
    2
    """
    return sum(map(_issafe1, _parse(data)))


def part2(data) -> int:
    """
    >>> part2(SAMPLE_INPUT)
    4
    """
    return sum(map(_issafe2, _parse(data)))


parts = (part1, part2)

File: rs/src/day2.rs
fn parse(data: &str) -> Vec<Vec<i32>> {
    data.lines()
        .flat_map(|line| {
            line.split_ascii_whitespace()
                .map(|level| level.parse::<i32>().ok())
                .collect::<Option<Vec<_>>>()
        })
        .collect::<Vec<_>>()
}

fn is_safe_1(report: &[i32]) -> bool {
    let mut decreasing = false;
    let mut increasing = false;
    report
        .iter()
        .zip(report.iter().skip(1))
        .all(|(x, y)| match x - y {
            -3..=-1 => {
                decreasing = true;
                !increasing
            }
            1..=3 => {
                increasing = true;
                !decreasing
            }
            _ => false,
        })
}

fn is_safe_2(report: &[i32]) -> bool {
    if report.is_empty() {
        return true;
    }
    let mut report2 = vec![0; report.len() - 1];
    report2.copy_from_slice(&report[1..]);
    is_safe_1(&report2)
        || (0..report2.len()).any(|i| {
            report2[i] = report[i];
            is_safe_1(&report2)
        })
}

pub fn part1(data: &str) -> usize {
    parse(data)
        .into_iter()
        .filter(|report| is_safe_1(report))
        .count()
}

pub fn part2(data: &str) -> usize {
    parse(data)
        .into_iter()
        .filter(|report| is_safe_2(report))
        .count()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        7 6 4 2 1
        1 2 7 8 9
        9 7 6 2 1
        1 3 2 4 5
        8 6 4 4 1
        1 3 6 7 9
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(2, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(4, part2(EXAMPLE));
    }
}
